### PR TITLE
added style guide for strong and weak self references

### PR DESCRIPTION
**Pull request by aboedo**
added sample code for capturing self weakly and strongly, and link to guide for when to do one or the other. 
Inspired by [this comment](https://github.com/RevenueCat/purchases-ios/pull/705#discussion_r683580830)

### CODE REVIEW

File: SwiftStyleGuide.swift [hunk 1]
```diff
@@ -104,6 +104,24 @@ private extension MyStruct {
...
+        // https://medium.com/flawless-app-stories/you-dont-always-need-weak-self-a778bec505ef
+
+        // no need to explicitly capture self if you need a strong reference
+        foo.methodThatNeedsStrongCapture() {
+            // ...
+        }
+
+        // but of course we do add it for weak references
+        foo.methodThatNeedsWeakCapture() { [weak self] in
+            guard let `self` = self else { return }
```
Comment (aboedo) [inline SwiftStyleGuide.swift:118]: I _think_ I recall seeing somewhere that the backticks aren't needed anymore, but I'm not sure

Comment (Juanpe) [reply SwiftStyleGuide.swift:118]: yep, you're right, the backticks are not needed anymore, `guard let self = self else { ... }` works fine 👍🏼

Comment (aboedo) [reply SwiftStyleGuide.swift:118]: awesome, thanks! I'll update accordingly

**Review by taquitos — APPROVED** [review]
Love it. One concern I have is linking to `medium.com`. If we can find the same resource but on github/any other non-pay-walled source, that would be ideal. I wouldn't block merging on this, but just something to think about.

**Comment by aboedo** [comment]
agreed. I'm actually surprised to realize it was on medium at all - I usually can't read anything from medium since I'm not subscribed to it. I'll try to find another source

**Comment by Juanpe** [comment]
Hey @aboedo, would it make sense to use Apple's doc link ([**Strong Reference Cycles for Closures**](https://docs.swift.org/swift-book/LanguageGuide/AutomaticReferenceCounting.html#ID56)) instead?

**Comment by aboedo** [comment]
@Juanpe good idea! I'll link to that one, thanks!

**Pull request merged**